presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java

```diff
@@ -347,7 +364,12 @@ public FunctionInfo resolveFunction(QualifiedName name, List<TypeSignature> para
             Map<String, Type> boundTypeParameters = function.getSignature().bindTypeParameters(resolvedTypes, false, typeManager);
             if (boundTypeParameters != null) {
                 checkArgument(match == null, "Ambiguous call to %s with parameters %s", name, parameterTypes);
-                match = function.specialize(boundTypeParameters, resolvedTypes.size(), typeManager, this);
+                try {
+                    match = specializedFunctionCache.get(new SpecializedFunctionKey(function, boundTypeParameters, resolvedTypes.size()));
+                }
+                catch (Exception e) {
```

Comment:
Assuming you use `getUnchecked`, this should be

```
catch (ExecutionException | UncheckedExecutionException e) {
    throw Throwables.propagate(e.getCause());
```

The `getRootCause()` method digs too deep